### PR TITLE
[te] Introduce likely/unlikely CompareSelect hint

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51771 [te] Benchmark vml-based logit
* #51752 [te] Implement log approximation using the VML approach
* **#51751 [te] Introduce likely/unlikely CompareSelect hint**
* #51749 [te][llvm] Generate arithmetic vs logical right shift as appropriate

Similar in spirit to the `__builtin_expect` C intrinsic, it's useful
to be able to hint the expected branch direction in a tensor expression.  Using
this flag has a few effects on codegen:

- The CompareSelect is generated using conditional branches, rather than selects
- The conditional branches are strongly hinted (like, 100000:1) in the indicated direction
- A vectorized hinted CompareSelect computes its condition in parallel with a
  mask "reduction" (e.g. a bitcast from `<i1 x 8>` to `<i*>`).  In AVX terms
  this sequence might look like:
```
vpcmpgtd %ymm0, %ymm1, %ymm2
vmovmskps %ymm2, %eax
```

The motivating case for this addition is an attempt I'm making to replicate
fast transcendentals using tensor expressions.  Floating-point numbers have
lots of special cases (denormals, inf, nan) that need special handling, and
it's convenient to be able to punt that handling off to a slow path while
keeping the fast path nice and tight.

Differential Revision: [D26246401](https://our.internmc.facebook.com/intern/diff/D26246401/)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 1]
```diff
@@ -698,53 +720,93 @@ void LLVMCodeGenImpl::visit(const Min* v) {
...
-      case CompareSelectOperation::kLT:
-        cmp_ = irb_.CreateFCmpOLT(lhs, rhs);
-        break;
-      case CompareSelectOperation::kLE:
-        cmp_ = irb_.CreateFCmpOLE(lhs, rhs);
-        break;
-      default:
-        // TODO: change to a proper error report
-        throw std::runtime_error("invalid operator type");
+  auto genUnbiased = [this, v]() -> llvm::Value* {
```
Comment (asuhan) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:723]: Do these two (`genUnbiased` and `genBiased`) really need to be lambdas, wouldn't they be better as methods?

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 2]
```diff
@@ -698,53 +720,93 @@ void LLVMCodeGenImpl::visit(const Min* v) {
...
+      auto maskType = llvm::Type::getIntNTy(getContext(), lanes);
+      auto zero = llvm::ConstantInt::get(maskType, 0);
+      auto mask = irb_.CreateBitOrPointerCast(cmp, maskType);
+      cmp = irb_.CreateICmpNE(mask, zero);
+    }
+
+    auto then_block = llvm::BasicBlock::Create(getContext(), "then", fn_);
+    auto else_block = llvm::BasicBlock::Create(getContext(), "else", fn_);
+    auto end_block = llvm::BasicBlock::Create(getContext(), "block", fn_);
+    constexpr int32_t total_weight = 100000;
```
Comment (asuhan) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:780]: What's the significance of 100000 here, couldn't it be 10 or even 1?

**Review by asuhan — APPROVED** [review]
Looks good, address the two comments if you agree.

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 1 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp:723]: I kind of prefer lambdas here for the maybe slightly silly reason that they want to use `this`, but they're just local helpers so I hate adding them to the class signature  ¯\_(ツ)_/¯

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 2 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp:780]: We talked about this offline, but just for posterity: this feature doesn't seem well documented.  I basically swiped this usage from the llvm repo itself, e.g. [this instrumentation in the ASAN implementation](https://github.com/llvm/llvm-project/blob/cfeecdf7b6df9cd89488948b440f8deeb458104c/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp#L1677).

**Pull request closed without merging**